nautobot/core/models/__init__.py

```diff
@@ -50,3 +97,92 @@ def validated_save(self):
         """
         self.full_clean()
         self.save()
+
+    def natural_key(self):
+        """
+        Smarter default implementation of natural key construction.
+
+        1. Handles nullable foreign keys (https://github.com/wq/django-natural-keys/issues/18)
+        2. Handles variadic natural-keys (e.g. Location model - [name, parent__name, parent__parent__name, ...].)
+        """
+        vals = []
+        for name_list in [name.split("__") for name in self.get_natural_key_fields()]:
+            val = self
+            for attr_name in name_list:
+                val = getattr(val, attr_name)
+                if val is None:
+                    break
+            vals.append(val)
+        # Strip trailing Nones from vals
+        cleaned_vals = []
+        for val in reversed(vals):
+            if val is None and not cleaned_vals:
+                continue
+            cleaned_vals = [val] + cleaned_vals
+        return cleaned_vals
+
+    natural_key_separator = "/"
```

Comment:
As Gary points out, per https://www.rfc-editor.org/rfc/rfc2396 there are several other characters we might consider, such as `~` or `!`

Reply:
We need to be cautious about whether these values would be permitted in fields for which uniqueness would be considered, for example `Device.name` or `Interface.name`, etc.

Reply:
I'm exploring changing this to `&` in an upcoming commit, with the reasoning as follows:

```python
# For our purposes, NATURAL_KEY_SLUG_SEPARATOR needs to be:
# 1. Safe in a URL path component (so that we can do URLS like "/dcim/devices/<natural_key_slug>/delete/")
#    Per RFC3986 section 2.3 the general "unreserved" characters are ALPHA and DIGIT and the characters -._~
#    Per RFC3986 section 3.3 path components also permit characters :@ and the "sub-delims" characters !$&'()*+,;=
# 2. Not readily confused with characters commonly seen in un-escaped natural key component fields
#    "." is already ruled out as an unreserved character but also would appear in IPv4 IPAddress and Prefix objects
#    ":" similarly would appear in IPv6 IPAddress/Prefix objects
#    "/" would appear in Prefix objects as well as various numbered device component names
```